pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -0,0 +1,260 @@
+func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling NodeMaintenance")
+	maintanenceMode := true
+
+	// Fetch the NodeMaintenance instance
+	instance := &kubevirtv1alpha1.NodeMaintenance{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			reqLogger.Info(fmt.Sprintf("NodeMaintenance Object: %s Deleted ", request.NamespacedName))
+			maintanenceMode = false
+		} else {
+			// Error reading the object - requeue the request.
+			reqLogger.Info("Error reading the request object, requeuing.")
+			return reconcile.Result{}, err
+		}
+	}
+
+	nodeName := request.Name
+
+	if maintanenceMode {
+		reqLogger.Info(fmt.Sprintf("Applying Maintenance mode on Node: %s with Reason: %s", nodeName, instance.Spec.Reason))
+	}
+	node, err := r.fetchNode(nodeName)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := runCordonOrUncordon(r, node, maintanenceMode); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if maintanenceMode {
+
+		taintRetries = 3
+		err = r.taintNodeForLiveMigration(nodeName)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		stop := make(chan struct{})
+		defer close(stop)
+
+		if err = r.startPodInformer(node, stop); err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Failed to start pod informer"))
+			return reconcile.Result{}, err
+		}
+
+		reqLogger.Info(fmt.Sprintf("Evict all Pods from Node: %s", nodeName))
+		if err = drainPods(r, nodeName); err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		uncordonedNode, err := r.fetchNode(nodeName)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		updated, err := removeTaint(r.client, uncordonedNode)
+		if !updated {
+			reqLogger.Error(err, fmt.Sprintf("kubevirt.io/drain taint was not removed from Node: %s", nodeName))
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileNodeMaintenance) fetchNode(nodeName string) (*corev1.Node, error) {
+	node := &corev1.Node{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node)
+	if err != nil && errors.IsNotFound(err) {
+		log.Error(err, fmt.Sprintf("Node: %s cannot be found", nodeName))
+		return nil, err
+	} else if err != nil {
+		log.Error(err, fmt.Sprintf("Failed to get Node %s: %v\n", nodeName, err))
+		return nil, err
+	}
+	return node, nil
+}
+
+func (r *ReconcileNodeMaintenance) startPodInformer(node *corev1.Node, stop <-chan struct{}) error {
+	fieldSelector := fields.SelectorFromSet(fields.Set{"spec.nodeName": node.Name})
```

Comment:
You will also have to select based on phases, you only want pods which are still running.

Reply:
ah nonsense.